src/main/java/org/dmfs/jems/optional/adapters/SingleOptional.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * Copyright 2018 dmfs GmbH
+ *
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.dmfs.jems.optional.adapters;
+
+import org.dmfs.jems.single.Single;
+import org.dmfs.jems.single.elementary.Frozen;
+import org.dmfs.optional.Optional;
+
+import java.util.NoSuchElementException;
+
+
+/**
+ * {@link Optional} that takes a {@link Single} of {@link Optional} and caches its value upon usage.
+ *
+ * @author Gabor Keszthelyi
+ */
+public final class SingleOptional<T> implements Optional<T>
+{
+    private Single<Optional<T>> mSingleOptional;
+
+
+    public SingleOptional(Single<Optional<T>> singleOptional)
+    {
+        mSingleOptional = new Frozen<>(singleOptional);
```

Comment:
Why the `Frozen`?

Reply:
To cache the value of the `Single`. In practice, the `Single` received there would almost always be a factory lambda/function which would be called twice when using the result `Optional`.